### PR TITLE
Add NavigatorExperimental mock

### PR DESCRIPTION
**Pull request by p16**
@lelandrichardson @RealOrangeOne I saw the discussion https://github.com/lelandrichardson/react-native-mock/pull/34#issuecomment-217684066 , is this a better way to implement a NavigationExperimental mock?

### CODE REVIEW

**Comment by RealOrangeOne** [comment]
This is definatlely a far better implementation of a mock over #34, but not all the keys of `navigatorExperimental` are components ([see here](https://github.com/facebook/react-native/blob/master/Libraries/NavigationExperimental/NavigationExperimental.js)). After having a brief look, only 2 of the keys are infact components. The source for all the keys can be found [here](https://github.com/facebook/react-native/tree/master/Libraries/NavigationExperimental). If you could add some mocks for these, that are similar to the original, that'd be great!

**Comment by p16** [comment]
@RealOrangeOne I'll have a look at them and update the PR

**Comment by RealOrangeOne** [comment]
Please update your branch from master, tests should now have been fixed

**Comment by sibelius** [comment]
👍

**Comment by p16** [comment]
@RealOrangeOne I've started my work on the components/modules.

I think we have two ways of mocking `NavigationExperimental`:
1. we mock the `NavigationExperimental` so that its "children" are just empty interfaces 
2. or we use the real one and mock only the base dependencies/components (`Animated`, `View`, etc.)

I've no strong opinion on which way could be better, but my feeling is that the first one can be the most pragmatic one and it will leave the door open to add behaviours to the mocks if a real need arise.

Do you have any thought/opinion on this?

/cc @lelandrichardson

**Comment by RealOrangeOne** [comment]
Personally, i'd go all out. Seeing as we're making the mock, we might as well make it as detailed as possible. `NavigatorExperimental` is an object, so lets keep it that way. If we can mock all it's children to be as simple as possible, but still roughly mimic the real objects.

File: src/react-native.js [hunk 1]
```diff
@@ -96,6 +96,8 @@ const ReactNative = {
   ColorPropType: require('./propTypes/ColorPropType'),
   EdgeInsetsPropType: require('./propTypes/EdgeInsetsPropType'),
   PointPropType: require('./propTypes/PointPropType'),
+  PointPropType: require('./propTypes/PointPropType'),
```
Comment (evollu) [inline src/react-native.js]: this is a duplicate?

Comment (p16) [inline src/react-native.js]: @evollu yep, I'll remove it. Good catch, thanks!

**Comment by p16** [comment]
@RealOrangeOne ok

**Comment by benjick** [comment]
Will this be merged soon or should I just use this PR? Cheers!

**Comment by RealOrangeOne** [comment]
@benjick  This can't be merged until the PR is complete. That's the point of the PR. It hasnt fixed the problem yet

**Comment by sibelius** [comment]
@benjick if u wanna test react-native-router-flux you can use mockery (http://stackoverflow.com/questions/37251534/how-to-test-project-when-3rd-party-modules-used-arent-transpiled/37655424#37655424)

**Comment by p16** [comment]
@RealOrangeOne I think the reducers have been removed 😁 less work for me! 😄  https://github.com/facebook/react-native/commit/3a8b50ad550b176bbce5713c5c74e1318c10f2a0

**Comment by RealOrangeOne** [comment]
@p16 Is that released into react-native yet? Not sure what other components there are that need mocking?

**Comment by p16** [comment]
@RealOrangeOne that's master. I did not find the commit in 0.27 nor on 0.28-rc0

**Comment by p16** [comment]
@RealOrangeOne I've basically copied `NavigationStateUtils` from the real one and remove all the external dependencies. But I've kept the logic in the methods because those can be used in react-native applications and the people using them would expect them to behave correctly. The same goes for `NavigationPropTypes`. 

Let me know what you think, I'm still wondering it this is the best approach...

/cc @lelandrichardson

**Comment by p16** [comment]
@RealOrangeOne can you have a look at `NavigationPropTypes` and `NavigationStateUtils`? I think they should be fine now.

I've mocked all the other components with `createMockComponent` because I could not find a good reason to unit test one of my components inside an `AnimatedView` or a `Transitioner`. Let me know if this is completely off or it make sense :)

File: src/Libraries/NavigationExperimental/index.js [hunk 2]
```diff
@@ -0,0 +1,19 @@
+/**
+ * @see  https://github.com/facebook/react-native/blob/master/Libraries/NavigationExperimental/NavigationExperimental.js
+ */
+import createMockComponent from './../../components/createMockComponent';
```
Comment (RealOrangeOne) [inline src/Libraries/NavigationExperimental/index.js]: you don't need the leading `./` on this path, you can start straight in with `../`

**Comment by RealOrangeOne** [comment]
From what I'm seeing, this mock looks almost perfect! 1 problem, which is something i''m hoping to address with the rest of the library, could you convert the `NavigatorExperimental` stuff to use ES6 syntax? I'm not liking the use of both `import` and `module.export` throughout this library, and am hoping to change it to completely ES6/7, the same as `react-native`. Everything is already set up to accomodate for this. Thanks

**Comment by p16** [comment]
@RealOrangeOne done both the comment and the ES6 syntax :)

**Comment by p16** [comment]
ups... forgot `NavigationExperimental/index.js`... fixing it

**Comment by RealOrangeOne** [comment]
Thanks for such a speedy response! Could you use the syntax that exports functions and vars in their definition, so we don't need the export block at the end of the file?
`export const var = '';`
`export function doThings()`

**Comment by p16** [comment]
@RealOrangeOne I should have removed the extra functions/constants

**Comment by p16** [comment]
@RealOrangeOne any news on this?

**Comment by lelandrichardson** [comment]
@p16 this looks like awesome work. Thanks!

I'm not sure if @RealOrangeOne's comment had this intention or not, but I believe where you are exporting `default` like here:

```
export default {
  foo,
  bar,
  ...
};
```

might be better off exported as:

```
module.exports = {
  foo,
  bar,
  ...
};
```

The reason is:

```
import { foo, bar } from 'baz';
```

and

```
const { foo, bar } = require('baz');
```

will only work in both cases if the `module.exports` syntax is used. Further, this is the syntax that react native uses, so it's best if our exported module has exactly the same shape.

Thus, if you just change `export default` to `module.exports =` in these cases, I believe we will be good to go!

**Comment by RealOrangeOne** [comment]
@lelandrichardson Thanks, I did not realise about this backwards compatibility issue! Yes once that's changed I think we can get this merged in!

**Comment by p16** [comment]
Ops, I got @RealOrangeOne 's comment completely wrong! Thanks @lelandrichardson , I'll fix the exports today.

**Comment by p16** [comment]
@RealOrangeOne @lelandrichardson done, sorry for the misunderstanding! :)

**Pull request merged**